modules/core/src/main/java/com/illposed/osc/transport/udp/OSCPortIn.java

```diff
@@ -59,73 +62,131 @@
 	private boolean resilient;
 	private Thread listeningThread;
 	private final OSCParserFactory parserFactory;
-	private final OSCPacketDispatcher dispatcher;
+	private final List<OSCPacketListener> packetListeners;
 
 	/**
 	 * Create an OSC-Port that listens on the given local socket for packets from {@code remote},
 	 * using a parser created with the given factory.
 	 * @param parserFactory to create the internal parser from
-	 * @param dispatcher to dispatch received and serialized OSC packets
+	 * @param packetListeners to handle received and serialized OSC packets
 	 * @param local address to listen on
 	 * @param remote address to listen to
 	 * @throws IOException if we fail to bind a channel to the local address
 	 */
 	public OSCPortIn(
 			final OSCParserFactory parserFactory,
-			final OSCPacketDispatcher dispatcher,
+			final List<OSCPacketListener> packetListeners,
 			final SocketAddress local,
 			final SocketAddress remote)
 			throws IOException
 	{
 		super(local, remote);
 
-		this.listening = false;
-		this.daemonListener = true;
-		this.resilient = false;
-		this.parserFactory = parserFactory;
-		if (dispatcher == null) {
-			this.dispatcher = new OSCPacketDispatcher();
-			// HACK We do this, even though it is against the OSC (1.0) specification,
-			//   because this is how it worked in this library until Feb. 2015.,
-			//   and thus users of this library expect this behaviour by default.
-			this.dispatcher.setAlwaysDispatchingImmediately(true);
-		} else {
-			this.dispatcher = dispatcher;
-		}
+		this.listening       = false;
+		this.daemonListener  = true;
+		this.resilient       = false;
+		this.parserFactory   = parserFactory;
+		this.packetListeners = packetListeners;
 	}
 
 	public OSCPortIn(
 			final OSCParserFactory parserFactory,
-			final OSCPacketDispatcher dispatcher,
+			final List<OSCPacketListener> packetListeners,
 			final SocketAddress local)
 			throws IOException
 	{
-		this(parserFactory, dispatcher, local, new InetSocketAddress(
-				OSCPort.generateWildcard(local), 0));
+		this(
+			parserFactory,
+			packetListeners,
+			local,
+			new InetSocketAddress(OSCPort.generateWildcard(local), 0)
+		);
 	}
 
-	public OSCPortIn(final OSCParserFactory parserFactory, final SocketAddress local)
-			throws IOException
+	private static List<OSCPacketListener> packetListeners(
+		final boolean defaultListener
```

Comment:
hmmm...
It fels somehow strange that some ctor's have `final List<OSCPacketListener> packetListeners` and others have `final boolean defaultListener`. I am not into the code (have not coded on the lib since some time), so maybe that would be problematic for some reason, but if not... it owuld be nice to have only one of the two, or none at all (so the default listener would have to be added manually) ... not sure...

how did you come up wiht this solution?

Reply:
I'm glad you mentioned this. It feels messy for sure. I'd like to use the Builder pattern to clean up the way OSCPortIn instances are created. I'm almost done taking an initial stab at that -- will update the PR soon.

Reply:
PR updated. I refactored a bit and implemented an OSCPortIn.Builder class. I think this makes the API much cleaner.